### PR TITLE
forc-crypto: vanity address generation

### PR DESCRIPTION
**Pull request by zees-dev**
## Description
Support for Bech32 case-insensitive vanity address generation via `forc-crypto` CLI tool.

Usage:
```sh
forc-crypto vanity --help
```

CLI options:
```
Generate a vanity address

Usage: forc-crypto vanity [OPTIONS]

Options:
      --starts-with <HEX_STRING>  Desired hex string prefix for the address
      --ends-with <HEX_STRING>    Desired hex string suffix for the address
      --regex <PATTERN>           Desired regex pattern to match the entire address (case-insensitive)
      --timeout <SECONDS>         Timeout in seconds for address generation
      --mnemonic                  Return mnemonic with address (default false)
      --save-path <PATH>          Path to save the generated vanity address to
  -h, --help                      Print help
  -V, --version                   Print version

Generate vanity addresses for the Fuel blockchain
```

## Example outputs
```sh
# > forc-crypto vanity --starts-with 0 --ends-with F

Starting to generate vanity address...
Successfully found vanity address in 0.012 seconds.
Address: 0d3c399e756dee9f7312215882f92685fdae25449bc74f33c31063000d68afdf
PrivateKey: cec536d4f5aae64685856ad36818777ba0aed349bdef848b487d6ebb1cc2a0a4
```

```sh
# > forc-crypto vanity --starts-with 0 --ends-with F --mnemonic

Starting to generate vanity address...
Successfully found vanity address in 0.141 seconds.
Address: 030b8484305c9f3af7b662d9fdd88dc75bdceb29f42d0f5ea5f72d3dfaf9380f
Mnemonic: found broccoli trap left thought attack quality smooth patrol enrich fault flavor legend amused monitor shoulder legend blast elbow custom dirt cotton tackle much
PrivateKey: 269ee26dd810a4a2df72969ed8941fe92ddc0bab236715535b867448c3ffa25e
```

## Relevant issues
- https://github.com/FuelLabs/sway/issues/6664

### CODE REVIEW

**Review by kayagokalp — CHANGES_REQUESTED** [review]
Hey @zees-dev, Nice job on your first PR!

I will be leaving a second review that actually checks the code but to start with can you:

1. run `cargo clippy`
2. format the `forc-crypto` `Cargo.toml` file, basically the order of dependencies needs to be alphabetical
3. Ensure that root Cargo.lock file is free of conflicts.

**Comment by kayagokalp** [comment]
Also if you use closes #6684. Github automatically marks the relevant issue  to be closed after this PR is merged.

File: forc-plugins/forc-crypto/src/keys/vanity.rs
Comment (JoshuaBatty) [inline forc-plugins/forc-crypto/src/keys/vanity.rs:38]: perhaps a timeout option could be nice? 

```rust
#[arg(long, value_name = "SECONDS")]
pub timeout: Option<u64>,
```

File: forc-plugins/forc-crypto/src/keys/vanity.rs
Comment (JoshuaBatty) [inline forc-plugins/forc-crypto/src/keys/vanity.rs:275]: Perhaps we could validate the the pattern length is no longer than 64

```rust
if pattern.len() > 64 { 
    return Err(anyhow!("Pattern too long: max 64 characters"));
}
````

File: forc-plugins/forc-crypto/src/keys/vanity.rs
Comment (JoshuaBatty) [inline forc-plugins/forc-crypto/src/keys/vanity.rs:146]: What are your thoughts about having some user feedback? something like...

```rust
pub fn find_vanity_address<T: VanityMatcher>(
    matcher: T, 
    use_mnemonic: bool,
) -> Result<(Address, SecretKey, Option<String>)> {
    let start = Instant::now();
    let mut attempts = AtomicUsize::new(0); // Need atomic for parallel iteration
    
    let result = wallet_generator(use_mnemonic)
        .for_each(|_| {
            let current = attempts.fetch_add(1, Ordering::Relaxed);
            if current % 10000 == 0 { // Pick a nice number that makes sense here, just going with 10000 
                let elapsed = start.elapsed().as_secs_f64();
                let rate = current as f64 / elapsed;
                println!("Tried {} addresses ({:.2} addresses/sec)...", current, rate);
            }
        })
        .find_any(|result| {
            if let Ok((addr, _, _)) = result {
                matcher.is_match(addr)
            } else {
                false
            }
        })
        .ok_or_else(|| anyhow!("Failed to generate a matching address"))?;

    Ok(result?)
}
```

**Review by JoshuaBatty — COMMENTED** [review]
Just a couple of suggestions. Great to see the benchmarking and tests added for this.

File: forc-plugins/forc-crypto/src/keys/vanity.rs
Comment (zees-dev) [reply forc-plugins/forc-crypto/src/keys/vanity.rs:146]: this can be a good idea; the `solana-keygen` seems to do this.

File: forc-plugins/forc-crypto/src/keys/vanity.rs
Comment (alfiedotwtf) [inline forc-plugins/forc-crypto/src/keys/vanity.rs:288]: From here we know it's a regex. Is there any need to touch it futher? i.e if the user is dropping down to a regex should we just let them match for whatever they're looking for and from here just pass it through?

File: forc-plugins/forc-crypto/src/keys/vanity.rs
Comment (alfiedotwtf) [inline forc-plugins/forc-crypto/src/keys/vanity.rs:68]: If the user is using a regex, should we let them match whatever they're looking for rather than only applying it to the start or end? i.e should this instead handle:

- left
- right
- left and right
- regex

The reason I'm asking is that if they're using a regex, they could put the left or right or both IN the regex.

**Comment by alfiedotwtf** [comment]
Nice!

Quick question - is there a big difference in timings between random vs mnemonic? I'm thinking it would always be a good idea to provide the mnemonic by default and if they're close in timing we should only provide that pathway... but if the performance of random is way faster, then yeah let's keep both.

**Comment by zees-dev** [comment]
> Nice!
> 
> Quick question - is there a big difference in timings between random vs mnemonic? I'm thinking it would always be a good idea to provide the mnemonic by default and if they're close in timing we should only provide that pathway... but if the performance of random is way faster, then yeah let's keep both.

There is a difference in timing; quite significant actually; here is benchmark:
```
#> cargo bench -p forc-crypto

     Running benches/bench_main.rs (target/release/deps/bench_main-77704bc04ac1c0dc)
Gnuplot not found, using plotters backend
Vanity Address Generation/Vanity Address (first byte 0xff)
                        time:   [656.07 µs 689.08 µs 731.09 µs]
                        change: [-6.1118% -0.1898% +7.1988%] (p = 0.96 > 0.05)
                        No change in performance detected.
Found 8 outliers among 100 measurements (8.00%)
  2 (2.00%) low mild
  1 (1.00%) high mild
  5 (5.00%) high severe
Benchmarking Vanity Address Generation/Vanity Address with Mnemonic (first byte 0xff): Warming up for 3.0000 s
Warning: Unable to complete 100 samples in 5.0s. You may wish to increase target time to 5.6s, or reduce sample count to 80.
Vanity Address Generation/Vanity Address with Mnemonic (first byte 0xff)
                        time:   [46.216 ms 55.127 ms 64.789 ms]
                        change: [-0.4164% +22.910% +51.685%] (p = 0.07 > 0.05)
                        No change in performance detected.
Found 2 outliers among 100 measurements (2.00%)
  1 (1.00%) high mild
  1 (1.00%) high severe
  ```

File: forc-plugins/forc-crypto/src/keys/vanity.rs
Comment (zees-dev) [reply forc-plugins/forc-crypto/src/keys/vanity.rs:68]: That's correct. However, in most cases, users typically choose a prefix and/or suffix based on characters, as this is what wallet UIs commonly display.
For a better user experience, I believe it's best to offer prefix/suffix options, since users are less likely to remember regex syntax offhand.

File: forc-plugins/forc-crypto/src/keys/vanity.rs
Comment (zees-dev) [reply forc-plugins/forc-crypto/src/keys/vanity.rs]: addressed: https://github.com/FuelLabs/sway/pull/6661/commits/e493891bc22b376d7175321e1367b2342587896f

File: forc-plugins/forc-crypto/src/keys/vanity.rs
Comment (zees-dev) [reply forc-plugins/forc-crypto/src/keys/vanity.rs]: addressed: https://github.com/FuelLabs/sway/pull/6661/commits/b16c926bada977a31c774bbd9197f969f3e8c516

File: forc-plugins/forc-crypto/src/keys/vanity.rs
Comment (zees-dev) [reply forc-plugins/forc-crypto/src/keys/vanity.rs]: addressed: https://github.com/FuelLabs/sway/pull/6661/commits/723db3ee288d309baab22f151b656246c955bb27

File: forc-plugins/forc-crypto/src/keys/vanity.rs
Comment (sdankel) [inline forc-plugins/forc-crypto/src/keys/vanity.rs:345]: We could put this in a clap validator function for the arguments, so the error comes from clap, as well as the length validation.

File: forc-plugins/forc-crypto/src/keys/vanity.rs
Comment (sdankel) [inline forc-plugins/forc-crypto/src/keys/vanity.rs:372]: We use the `tracing` library in forc and forc plugins to make it easier to silence output or provide verbose output in the CLIs. These `write!`s can be replaced with `tracing::error!`

File: forc-plugins/forc-crypto/src/keys/vanity.rs
Comment (sdankel) [inline forc-plugins/forc-crypto/src/keys/vanity.rs:164]: If you want to make it look extra cool, you could use [indicatif](https://github.com/console-rs/indicatif/tree/main) and make this statement update the address count and rate update on the same line, similar to [this example](https://github.com/console-rs/indicatif/blob/main/examples/yarnish.rs#L84)

File: forc-plugins/forc-crypto/src/keys/vanity.rs
Comment (sdankel) [inline forc-plugins/forc-crypto/src/keys/vanity.rs:344]: Since we aren't supporting other regex functionality like globs, perhaps we should only support string matching (without '^' or '$') and update the help text?

File: forc-plugins/forc-crypto/src/keys/vanity.rs
Comment (sdankel) [inline forc-plugins/forc-crypto/src/keys/vanity.rs:30]: Related to my other comment, since we aren't really supporting regex I think this should be:

```suggestion
    /// Desired hex sting prefix.
    #[arg(long, value_name = "HEX_STRING", required_unless_present = "ends_with")]
    pub starts_with: Option<String>,
```

And the same for ends_with.

File: forc-plugins/forc-crypto/src/lib.rs
Comment (sdankel) [inline forc-plugins/forc-crypto/src/lib.rs:7]: The other forc plugins in this repo use the `forc_util::cli_examples` macro. It might be good to use that here for consistency.

File: forc-plugins/forc-crypto/src/keys/vanity.rs
Comment (sdankel) [inline forc-plugins/forc-crypto/src/keys/vanity.rs:113]: nit: I think this should be a separate error, `InvalidPatternCombination`, since it is different and it will make the error output a little cleaner.

File: forc-plugins/forc-crypto/src/keys/vanity.rs
Comment (JoshuaBatty) [reply forc-plugins/forc-crypto/src/keys/vanity.rs:164]: Oh nice, indicatif looks really cool!

File: forc-plugins/forc-crypto/src/keys/vanity.rs
Comment (zees-dev) [reply forc-plugins/forc-crypto/src/keys/vanity.rs]: somewhat addressed by the refactor: https://github.com/FuelLabs/sway/pull/6661/commits/59ab719069be67c3a2ef209e632c2d5039996746

File: forc-plugins/forc-crypto/src/keys/vanity.rs
Comment (zees-dev) [reply forc-plugins/forc-crypto/src/keys/vanity.rs]: addressed: https://github.com/FuelLabs/sway/pull/6661/commits/59ab719069be67c3a2ef209e632c2d5039996746

File: forc-plugins/forc-crypto/src/keys/vanity.rs
Comment (zees-dev) [reply forc-plugins/forc-crypto/src/keys/vanity.rs]: addressed: https://github.com/FuelLabs/sway/pull/6661/commits/6746af46c48b9873e9a649fee075e7147e6069fd

File: forc-plugins/forc-crypto/src/lib.rs
Comment (zees-dev) [reply forc-plugins/forc-crypto/src/lib.rs:7]: cannot find an example specifically for top-level help; also this code remains unchanged as it was migrated from `main.rs`.

File: forc-plugins/forc-crypto/src/keys/vanity.rs
Comment (zees-dev) [reply forc-plugins/forc-crypto/src/keys/vanity.rs:336]: addressed: https://github.com/FuelLabs/sway/pull/6661/commits/59ab719069be67c3a2ef209e632c2d5039996746

File: forc-plugins/forc-crypto/src/keys/vanity.rs
Comment (zees-dev) [reply forc-plugins/forc-crypto/src/keys/vanity.rs]: addressed: https://github.com/FuelLabs/sway/pull/6661/commits/80ca7a2ce4a87e0c8a6038ff69defc3e07af117d

File: forc-plugins/forc-crypto/src/keys/vanity.rs
Comment (zees-dev) [reply forc-plugins/forc-crypto/src/keys/vanity.rs]: removed custom error as no longer needed as it did not see much usage; reverting back to anyhow: https://github.com/FuelLabs/sway/pull/6661/commits/80ca7a2ce4a87e0c8a6038ff69defc3e07af117d

File: forc-plugins/forc-crypto/src/keys/vanity.rs
Comment (zees-dev) [reply forc-plugins/forc-crypto/src/keys/vanity.rs]: removed custom error as no longer needed as it did not see much usage; reverting back to anyhow: https://github.com/FuelLabs/sway/commit/80ca7a2ce4a87e0c8a6038ff69defc3e07af117d

File: forc-plugins/forc-crypto/src/keys/vanity.rs
Comment (zees-dev) [reply forc-plugins/forc-crypto/src/keys/vanity.rs:226]: had a look; cant think of a great UI; also this dependency is not currently in the project.
Maybe a future enhancement.

File: Cargo.lock
Comment (sdankel) [inline Cargo.lock:2768]: Should this be a dev dependency?

File: forc-plugins/forc-crypto/src/keys/vanity.rs
Comment (alfiedotwtf) [inline forc-plugins/forc-crypto/src/keys/vanity.rs:39]: I don't think this is needed since you're going to return error if `Regex::new()` fails. This will also fail with character classes e.g `--regex "[[:alpha:]]+"`

**Review by alfiedotwtf — DISMISSED** [review]

File: forc-plugins/forc-crypto/src/keys/vanity.rs
Comment (zees-dev) [reply forc-plugins/forc-crypto/src/keys/vanity.rs]: addressed: https://github.com/FuelLabs/sway/pull/6661/commits/5a850d2f485b61a3ef3b1febd1432a9813005253

File: Cargo.lock
Comment (zees-dev) [reply Cargo.lock:2768]: yeap this is declared as dev dependency in `forc-crypto` `Cargo.toml`

**Review by sdankel — APPROVED** [review]
Looks good! :shipit:

**Review by kayagokalp — DISMISSED** [review]
Looks good to me!

**Comment by kayagokalp** [comment]
I think only thing left in CI is to fix the tests generated by `forc_util::cli_examples!` macro.

**Review by kayagokalp — APPROVED** [review]

**Review by JoshuaBatty — APPROVED** [review]
Great work!

**Pull request merged**
